### ISSUE TITLE
False positive "Objects are always passed by reference" when reference is actually needed

### ISSUE THREAD

**Issue by chylex** [kalessil/phpinspectionsea#1623]
| Subject | Details |
| :--- | :--- |
| Plugin | Php Inspections (EA Extended), version 4.0.5 |
| Language level | PHP 7.4 |

Bug description
--

The following minimal example gets `B::__construct` marked with "Objects are always passed by reference" and suggests replacing `A &$a_ref` with `A $a_ref`.

The actual intention is to give `B` ownership of the *reference itself*, so that it can reassign it. Applying the inspection removes that reference, causing `$a` to never get assigned so the code throws an error.

I'm not sure how common this pattern is in PHP code, but it should be enough to detect when a method parameter such as `&$a_ref` is used inside that method again as `&$a_ref` rather than `$a_ref`, and not show the intention in such case.

```php
class A{
  public function test(): void{
    echo 'hi';
  }
}

class B{
  private ?A $a_ref;
  
  public function __construct(?A &$a_ref){
    $this->a_ref = &$a_ref;
  }
  
  public function assign(): void{
    $this->a_ref = new A();
  }
}

$a = null;
$b = new B($a);
$b->assign();
$a->test();
```

**Comment by informatik-handwerk** [comment]
I am experiencing this as well.

# There is a misunderstanding over what a "passing by reference" is

It is about the variable - memory location, yes, having certain content.  
But it goes beyond this content, the value inside the variable.

## Passing by reference

You can change the content in the context of caller - are writing to the same memory location. When I "pass object by reference" to a callee, the callee can assign into that variable a string and I as the caller, have suddenly also a string there, instead of that object.

## Passing by value

New memory is allocated for the argument. The content of that memory location depends on the type:
- In case of primitives, the content is copied.
- In case of arrays (and possibly strings), afaiui, the new memory location points to a lazy cloner which clones the value upon mutation operation.
- In case of objects (also resources), the new memory location points to the same object.

# The classification of variable passing is more broad than by reference/by value

There are many modalities, but the simplest illustrating the point is:
1) a passing-type which exposes the memory location (in PHP: passing by reference)
2) a passing-type which allows mutation of the value (in PHP: passing of objects)
3) one where the value is (lazy-)detached. (in PHP: passing by value of primitives and arrays)

A couple of interesting points:
- **You cannot pass an array to have it changed by caller, without giving away the control of getting it overwritten altogether by anything.** You jump from 3 directly to 1 when switching from "by value" to "by reference".
- In the development of PHP, people did not attempt to plug in cloning of objects and resources into this passing-type split, which would make things more interesting.
- I am not sure how arrays with references inside them behave, I remember the references being actually kept, because the by value only applies to the carrier array.

**Comment by neuberkauf** [comment]
i'm seeing a similar problem with reference parameters, but not _exactly_ as described...

```php
// PHP 8.1
class Settings {
    public function __construct(private array $settings) {}

    public function get(string $key) {
        return $this->find($key, $this->settings);
    }

    public function set(string $key, mixed $newValue) {
        [$chain, $link] = explode('.', $key);
        $oldKey = &$this->find(implode('.', $chain), $this->settings);
        $oldKey[$link] = $newValue;
    }

    private function &find(string $key, Settings|array &$settings) {
        /*
         * Relevant functionality to show how the reference is used
         */
        [$link, $chain] = explode('.', $key, 2);

        if ($settings instanceof Settings) {
            $found = $settings->get($link);
        } elseif (is_array($settings) && array_key_exists($link, $settings)) {
            $found = &$settings[$link];
        }
        // Else other detection logic, ultimately end up with:

        if ($chain !== '') {
            /*
             * Recursive call is needed as there are still links in the chain to find
             */
            $found = &$this->find($chain, $found);
        }

        return $found;
    }
}
```
`get()` doesn't care about the reference, obviously. But `get()`/`set()` both use the same location logic, so `find()` returns a reference so `set()` can update a value.

However, internally the reference points to the _passed in_ reference to the starting point (or recursive reference).

The `Settings` class is capable of having nested `Settings` instances, so the finder has to allow it to be an argument.

All that is to set up where the inspection is giving a false positive:

```php
    private function &find(string $key, Settings|array &$settings) {
```
The `$settings` parameter _must_ be a reference when the argument is an array, but the inspection is ignoring the primitive type (which it normally wouldn't complain about) because i'm using a union type hint with a class.

Maybe this is the same issue, or should i file a new one?

**Comment by Benjaminhu** [comment]
| Subject | Detail |
|--------|--------|
| PhpStorm | 2024.1.6 |
| PHP | 8.2 |
| Build | #PS-241.19072.22, built on August 9, 2024 |
| Runtime version | 17.0.11+1-b1207.30 amd64 |
| Plugin | com.kalessil.phpStorm.phpInspectionsEA (5.1.0) |

+1 false positive:

![image](https://github.com/user-attachments/assets/c452a94d-d490-4778-8fed-888a50fd0f62)
